runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -0,0 +1,184 @@
+        Preconditions.checkArgument(Address.NON_ADDRESS == -1);
+        Preconditions.checkArgument(lastAddressRead >= Address.NON_ADDRESS,
+                "lastAddressRead %s must be >= %s", lastAddressRead, Address.NON_ADDRESS);
+        Preconditions.checkArgument(bufferSize > 0, "buffer size %s must be > 0", bufferSize);
+        this.streamId = streamId;
+        this.bufferSize = bufferSize;
+        this.addressesToRead = BufferUtils.synchronizedBuffer(new BoundedFifoBuffer(this.bufferSize));
+        this.lastAddressRead = lastAddressRead;
+        this.maxAddressSeen = this.lastAddressRead;
+        this.trimMark = Address.NON_ADDRESS;
+    }
+
+    /**
+     * Check if a the last address read was trimmed, since the last address has to always be
+     * greater than the trim mark, if this condition is trim, then a trim has created a gap
+     * that hasn't be synced and cannot be recovered.
+     */
+    private boolean trimGap() {
+        return trimMark != Address.NON_ADDRESS && lastAddressRead < trimMark;
+    }
+
+    //max has to always be greater than last address read
+    // called by the scheduler thread
+    public int availableSpace() {
+        int availableSpace = this.bufferSize - addressesToRead.size();
+        Preconditions.checkState(availableSpace >= 0);
+        return availableSpace;
+    }
+
+    public void refresh(StreamAddressSpace streamAddressSpace) {
+        if (log.isTraceEnabled()) {
+            log.trace("refresh id {} max {} trim {} newTm {} newAddresses {}", streamId, maxAddressSeen, trimMark,
+                    streamAddressSpace.getTrimMark(), streamAddressSpace.toArray());
+        }
+
+        if (lastAddressRead < streamAddressSpace.getTrimMark()
+                || maxAddressSeen < streamAddressSpace.getTrimMark()) {
+
+            // Since a trim has been detected, there is no needed to transfer the addresses, we just
+            // need to update the trim mark, on the subsequent sync, the task will error out before attempting
+            // to read, notice that a trim exception can occur even if streamaddressspace is empty
+            trimMark = streamAddressSpace.getTrimMark();
+
+            // if the task is syncing it will either discover the trim during reading the addresses that
+            // are already buffered, or discover this trim the next time the task gets scheduled
+
+            // since refresh can be called while the task is syncing we can error out when the scheduling tries to
+            // reschedule this task
+            return;
+        }
+
+        Preconditions.checkState(trimMark <= streamAddressSpace.getTrimMark(),
+                "streamId=%s new trimMark %s has to be >= %s", streamId, streamAddressSpace.getTrimMark(), trimMark);
+
+        // Adopt new trim mark
+        trimMark = streamAddressSpace.getTrimMark();
+
+        int maxAddressesToTransfer = availableSpace();
+        long[] newAddresses = streamAddressSpace.toArray();
```

Comment:
why not just use `streamAddressSpace` list? why you need to create an array (and consume 2x mem) from the list?